### PR TITLE
session/variable: forbid changing @@global.require_secure_transport to 'on' with SEM enabled

### LINKED ISSUE #47665: Add read-only restriction on require_secure_transport variable
## Enhancement
At present, [TiDB Dedicated](https://docs.pingcap.com/tidbcloud/select-cluster-tier#tidb-dedicated)  does not support this variable. It is necessary to add a read-only restriction for require_secure_transport to prevent unintentional enabling that could result in SQL client connection failures.

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #47665

Problem Summary:

The configuration for TiDB Cloud dedicated cluster is:

```
require_secure_transport: OFF
ssl_ca : /var/lib/tidb-server-tls/ca.crt
ssl_cert : /var/lib/tidb-server-tls/tls.crt
ssl_key: /var/lib/tidb-server-tls/tls.key
```

We just provide the `ca.crt` file for the the users to connect the TiDB cluster. When `require_secure_transport = OFF`, there is no problem.

**But if the user set the value of require_secure_transport to 'ON', the users can not connect to the cluster anymore!**

This is because the new configuration require the mysql client to pass all `ca.crt` `tls.crt` `tls.key` correctly, while the user only have  the `ca.crt` file.

### What is changed and how it works?

To prevent the user from losing access of the cluster, we can forbid changing `require_secure_transport` system variable to 'on' as a workaround, then the user can not modify it and trigger the issue.

### Check List

Tests 

- [X] Unit test
- [ ] Integration test
- [X] Manual test (add detailed scripts or steps below)

Modify the config.toml.example:
```
# Security Enhanced Mode (SEM) restricts the "SUPER" privilege and requires fine-grained privileges instead.
enable-sem = true
```
```
./bin/tidb-server -config ./pkg/config/config.toml.example
mysql -h 127.0.0.1 -u root -P 4000
```
```
mysql> set @@global.require_secure_transport = on;
ERROR 1105 (HY000): require_secure_transport can not be set to ON with SEM(security enhanced mode) enabled
```

- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [X] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Forbid changing @@global.require_secure_transport variable to 'on' when SEM(security enhanced mode) is enabled
```

### CODE REVIEW

File: pkg/sessionctx/variable/sysvar.go [hunk 1]
```diff
@@ -1117,6 +1117,14 @@ var defaultSysVars = []*SysVar{
...
+				// @@global.require_secure_transport=on, he can not login the cluster anymore!
+				// A workaround for this is making require_secure_transport read-only for that case.
+				// SEM(secure enhanced mode) is enabled by default with only that settings.
+				cfg := config.GetGlobalConfig()
+				if cfg.Security.EnableSEM {
+					return "", errors.New("require_secure_transport can not be set to ON with SEM(security enhanced mode) enabled")
+				}
 				// Refuse to set RequireSecureTransport to ON if the connection
 				// issuing the change is not secure. This helps reduce the chance of users being locked out.
 				if vars.TLSConnectionState == nil {
```
Comment (tiancaiamao) [inline pkg/sessionctx/variable/sysvar.go:1130]: The current PR is a workaround.
Maybe in a real fix we can make this check stronger? vars.TLSConnectionState can be non-nil even when tls.crt and tls.key are not provided.
If the check become tls.ca && tls.crt && tls.key, maybe things become better?

Comment (bb7133) [reply pkg/sessionctx/variable/sysvar.go:1130]: LGTM

File: pkg/sessionctx/variable/sysvar.go [hunk 2]
```diff
@@ -1117,6 +1117,14 @@ var defaultSysVars = []*SysVar{
 			return nil
 		}, Validation: func(vars *SessionVars, normalizedValue string, originalValue string, scope ScopeFlag) (string, error) {
 			if vars.StmtCtx.StmtType == "Set" && TiDBOptOn(normalizedValue) {
+				// On tidbcloud dedicated cluster with the default configuration, if an user modify
+				// @@global.require_secure_transport=on, he can not login the cluster anymore!
+				// A workaround for this is making require_secure_transport read-only for that case.
+				// SEM(secure enhanced mode) is enabled by default with only that settings.
```
Comment (bb7133) [inline pkg/sessionctx/variable/sysvar.go:1123]:
```suggestion
				// SEM(security enhanced mode) is enabled by default with only that settings.
```

**Review by bb7133 — APPROVED** [review]
LGTM

**Comment by bb7133** [comment]
I plan to collect 2 LGTMs to ensure this code change is ready.

**Comment by bb7133** [comment]
However, whether it should be merged into `master` branch is a question to me(after all it is just a workaround).

**Review by CbcWestwolf — APPROVED** [review]

**Pull request merged**